diag/diag.proto

```diff
@@ -0,0 +1,238 @@
+  // default invalid choice.
+  BERT_STATUS_UNKNOWN = 0;
+  // BERT requests (Start, Stop, GetStatus) were processed successfully.
+  BERT_STATUS_OK = 1;
+  // The specified port was not found.
+  BERT_STATUS_NON_EXISTENT_PORT = 2;
+  // HW error was encountered while performing BERT operation.
+  BERT_STATUS_HARDWARE_ACCESS_ERROR = 3;
+  // PRBS generating polynomial is not supported by the target.
+  BERT_STATUS_UNSUPPORTED_PRBS_POLYNOMIAL = 4;
+  // There is already a BERT running on the specified port. Returned when
+  // `StartBert` RPC tries to add run BERT on an already in-use port.
+  BERT_STATUS_PORT_ALREADY_IN_BERT = 5;
+  // There is no BERT running on the specified port. Returned when `StopBert`
+  // or `GetBertResult` RPC was called for an idle port.
+  BERT_STATUS_PORT_NOT_RUNNING_BERT = 6;
+  // The specified test duration is too small.
+  BERT_STATUS_TEST_DURATION_TOO_SHORT = 7;
+  // The specified test duration is larger than maximum allowed.
+  BERT_STATUS_TEST_DURATION_TOO_LONG = 8;
+  // The given BERT operation ID is not known. Returned for `StopBert` and
+  // `GetBertResult` RPCs.
+  BERT_STATUS_BERT_OPERATION_ID_NOT_FOUND = 9;
+  // The given BERT operation ID is already in use. Returned when `StartBert`
+  // RPC uses an ID which is already memorized for a BERT operation.
+  BERT_STATUS_BERT_OPERATION_ID_ALREADY_IN_USE = 10;
+  // Failure to get the peer lock.
+  BERT_STATUS_PEER_LOCK_FAILURE = 11;
+  // Lost the peer lock after locking once.
+  BERT_STATUS_PEER_LOCK_LOST = 12;
+  // Misc internal errors that cannot be categorized by any of the previous
+  // error codes.
+  BERT_STATUS_INTERNAL_ERROR = 13;
+}
+
+message StartBERTRequest {
+  // Per port BERT start requests.
+  message PerPortRequest {
+    // Path to the interface corresponding to the port.
+    gnoi.Path interface = 1;  // required
+    // The selected PRBS generating polynomial for BERT.
+    PrbsPolynomial prbs_polynomial = 2;  // required
+    // BERT duration in mins. Must be a positive number.
+    uint32 test_duration_in_mins = 3;  // required
+  }
+  // Unique BERT operation ID specified by the client. Multiple BERTs run on
+  // different ports can have the same BERT operation ID. This ID will be used
+  // later to stop the operation and/or get its results.
+  string bert_operation_id = 1;
+  // All the per-port BERTs that are considered one BERT operation and have the
+  // same BERT operation ID.
+  repeated PerPortRequest per_port_requests = 2;
+}
+
+message StartBERTResponse {
+  // Per-port BERT start responses.
+  message PerPortResponse {
+    // Path to the interface corresponding to the port.
+    gnoi.Path interface = 1;
+    // BERT start status for this port.
+    BertStatus status = 2;
+  }
+  // The same BERT operation ID given by the request.
+  string bert_operation_id = 1;
+  // Captures the results of starting BERT on a per-port basis.
+  repeated PerPortResponse per_port_responses = 2;
+}
+
+message StopBERTRequest {
```

Comment:
Here's where I think things get a bit more complicated by having multiple ports be part of the same BERT request. If I stop a BERT only on one port, but it has a single operation ID for that BERT, then this means that the BERT is cancelled for a subset of the ports, but running for the remainder. The current proto requires the caller to always check all statuses of the underlying ports to know the overall status. It seems like 1:1 would simplify this.

Reply:
I strongly believe the capability of running BERT on multiple port in one RPC is required for tools. As also explained above. 

Regarding your example: 
The user can start BERT on multiple ports and put them under the same operation ID. That is true. He/she can also stop BERT on one or more ports which were part of the same operation ID later. The success is determines based on the overall BER values captured on all the ports which we get via GetBERTResult RPC. So if some of the per port BERTs were stopped earlier, things should still be clear. The BERT results will tell how long BERT were run on those ports and what was the BER value on those. To me this seems clear.

Reply:
Couldn't we simplify the GetBERTResult by just having it return status for all the ports that were associated with the operation id (then don't need the two modes, all and per-port)

Reply:
So as I understand it, the tools have a use case to get the results for all the ports that might be part of different operation IDs.

As you said we might be able to simplify the logic a bit. I added a TODO there. Happy to simplify in case there is no use case.